File: OOPS_LAB/Program_2/date.java
class DATEFORMATE {
    private int day;
    private int month;
    private int year;

    public DATEFORMATE(int dd, int mm, int yyyy) {
        day = dd;
        month = mm;
        year = yyyy;
    }

    public boolean ValidateDate() {
        if (year < 1 || year > 9999 || month < 1 || month > 12 || day < 1 || day > 31) {
            return false;
        }

        if (month == 2) {
            boolean isLeap = (year % 4 == 0 && (year % 100 != 0 || year % 400 == 0));
            if (isLeap) {
                return (day <= 29);
            } else {
                return (day <= 28);
            }
        }

        if (month == 4 || month == 6 || month == 9 || month == 11) {
            return (day <= 30);
        }

        return true;
    }

    public void Display() {
        System.out.printf("%02d/%02d/%04d\n", day, month, year);
    }
}

public class date {
    public static void main(String[] args) {
        DATEFORMATE d1 = new DATEFORMATE(29, 2, 2023);
        
        if (d1.ValidateDate()) {
            System.out.print("Date is Valid: ");
            d1.Display();
        } else {
            System.out.println("Date is Invalid!");
        }
    }
}

File: OOPS_LAB/Program_2/date.cpp
#include<iostream>
using namespace std;

class DATE{
    private:
        int day;
        int month;
        int year;
    public:
        DATE()
        {
            day=0;
            month=0;
            year=0;
        }

        DATE(int dd, int mm, int yyyy)
        {
            day=dd;
            month=mm;
            year=yyyy;
        }

        bool ValidateDate()
        {
            if (this->day<=0 && this->day>31)
            {
                return false;
            }
            if (this->day<=0 && this->month>12)
            {
                return false;
            }
            if (this->year<=0 && this->year>9999)
            {
                return false;
            }
            
            if (this->month==2)
            {
                if (this->year%4==0)
                {
                    if (this->day>29)
                    {
                        return false;
                    }
                }
                else{if (this->day>28)
                    {
                        return false;
                    }}
            }
            return true;
        }

        void display()
        {
            cout<<this->day<<"/"<<this->month<<"/"<<this->year<<endl;
        }
};

int main(int argc, char const *argv[])
{
    DATE* d1 = new DATE(29,2,2001);
    if(d1->ValidateDate())
    {
        cout<<"\nDate is valid"<<endl;
        d1->display();
    }
    else{
        cout<<"\nDate is invalid";
    }
    return 0;
}
